rs/rosetta-api/ledger_core/src/approvals.rs

```diff
@@ -138,42 +252,35 @@ where
         self.check_postconditions();
         r
     }
-}
 
-impl<K, AccountId, Tokens> Approvals for AllowanceTable<K, AccountId, Tokens>
-where
-    K: Ord + for<'a> From<(&'a AccountId, &'a AccountId)> + Clone,
-    K: Into<(AccountId, AccountId)>,
-    AccountId: std::cmp::PartialEq,
-    Tokens: TokensType,
-{
-    type AccountId = AccountId;
-    type Tokens = Tokens;
-
-    fn allowance(
+    /// Returns the current spender's allowance for the account.
+    pub fn allowance(
         &self,
-        account: &AccountId,
-        spender: &AccountId,
+        account: &AD::AccountId,
+        spender: &AD::AccountId,
         now: TimeStamp,
-    ) -> Allowance<Tokens> {
-        let key = K::from((account, spender));
-        match self.allowances.get(&key) {
+    ) -> Allowance<AD::Tokens> {
+        match self
+            .allowances_data
+            .get_allowance(&(account.clone(), spender.clone()))
```

Comment:
These (and other) `clone()` calls caught my eye and I tried to remove them, since ideally I think a reference to the key should be enough for lookups, but it doesn't seem so straight-forward. I saw that there was also some related discussion in the earlier MR on GitLab. Did you already look into this? I found [this discussion](https://stackoverflow.com/questions/45786717/how-to-get-value-from-hashmap-with-two-keys-via-references-to-both-keys) on SO which seems related, but I don't know if that would apply to [`StableBTreeMap`](https://github.com/dfinity/stable-structures/blob/817a163e0f8a5a2afaad0469e9434398ac6bf7d3/src/btreemap.rs#L512-L521)s.

Did you run any (`canbench` or other) performance measurements to see if this change has any impact on the ledger performance? I wouldn't expect there to be much, since cloning `AccountId`s should be too expensive, but maybe this having something in place, and being able to compare the baseline with whatever changes we make going forward would be useful, WDYT?

I don't consider any of the above blocking for this MR or ticket - discussing and investigating further in a separate ticket would be fine with me!

Reply:
I measured the performance of all icrc2 methods in the ICRC and the ICP ledger. The increases in instruction usage are as follows:
ICRC:
approve: 0.2%
allowance: 0.34%
transfer_from: 0.43%
ICP:
approve: 0.65%
allowance: -0.01%
transfer_from: 2.86%
I will have a closer look at the last one!

Reply:
As suggested above and discussed offline, the increases are not blocking for this PR. I created a separate ticker for further investigation.